staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/admission.go

```diff
@@ -0,0 +1,106 @@
+
+// NewManagedFieldsValidatingAdmissionController validates the managedFields after calling
+// the provided admission and resets them to their original state if they got changed to an invalid value
+func NewManagedFieldsValidatingAdmissionController(wrap admission.Interface) admission.Interface {
+	return nil
+}
+
+type managedFieldsValidatingAdmissionController struct {
+	wrap admission.Interface
+}
+
+var _ admission.Interface = &managedFieldsValidatingAdmissionController{}
+var _ admission.MutationInterface = &managedFieldsValidatingAdmissionController{}
+var _ admission.ValidationInterface = &managedFieldsValidatingAdmissionController{}
+
+// Handles calls the wrapped admission.Interface if aplicable
+func (admit *managedFieldsValidatingAdmissionController) Handles(operation admission.Operation) bool {
+	return admit.wrap.Handles(operation)
+}
+
+// Admit calls the wrapped admission.Interface if aplicable and resets the managedFields to their state before admission if they
+// got modified in an invalid way
+func (admit *managedFieldsValidatingAdmissionController) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	mutationInterface, isMutationInterface := admit.wrap.(admission.MutationInterface)
+	if !isMutationInterface {
+		return nil
+	}
+	objectMeta, err := meta.Accessor(a.GetObject())
+	if err != nil {
+		return err
+	}
+	managedFieldsBeforeAdmission := objectMeta.GetManagedFields()
+	if err := mutationInterface.Admit(ctx, a, o); err != nil {
+		return err
+	}
+	objectMeta, err = meta.Accessor(a.GetObject())
+	if err != nil {
+		return err
+	}
+	managedFieldsAfterAdmission := objectMeta.GetManagedFields()
+	if err := validateManagedFields(managedFieldsAfterAdmission); err != nil {
+		objectMeta.SetManagedFields(managedFieldsBeforeAdmission)
+		warning.AddWarning(ctx, "",
+			fmt.Sprintf(InvalidManagedFieldsAfterMutatingAdmissionWarningFormat,
+				err.Error()),
+		)
+	}
+	return nil
+}
+
+// Validate calls the wrapped admission.Interface if aplicable
+func (admit *managedFieldsValidatingAdmissionController) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if validationInterface, isValidationInterface := admit.wrap.(admission.ValidationInterface); isValidationInterface {
+		return validationInterface.Validate(ctx, a, o)
+	}
+	return nil
+}
+
+func validateManagedFields(managedFields []metav1.ManagedFieldsEntry) error {
+	for i, managed := range managedFields {
+		if len(managed.APIVersion) < 1 {
+			return fmt.Errorf(".metadata.managedFields[%d]: missing apiVersion", i)
+		}
+		if len(managed.FieldsType) < 1 {
+			return fmt.Errorf(".metadata.managedFields[%d]: missing fieldsType", i)
+		}
+		if len(managed.Manager) < 1 {
+			return fmt.Errorf(".metadata.managedFields[%d]: missing manager", i)
+		}
+		if managed.FieldsV1 == nil {
+			return fmt.Errorf(".metadata.managedFields[%d]: missing fieldsV1", i)
+		}
+	}
```

Comment:
Can you remind me what we discussed about parsing the actual fieldset?

Reply:
We wanted to only validate that managedFields has the right format. I don't remember that we actually wanted to parse FieldsV1 here and don't think it's necessary.
The only way I can think of the contents of FieldsV1 is somebody actually modifying it in their webhook and that should be highly discouraged (especially because our parsing logic is /internal/ right?).
But now that I think about it, somebody will most certainly get that idea.